ouroboros-network-framework/test/Test/Ouroboros/Network/Server2.hs

```diff
@@ -2123,6 +2131,90 @@ prop_connection_manager_valid_transition_order serverAcc (ArbDataFlow dataFlow)
+prop_connection_manager_counters serverAcc (ArbDataFlow dataFlow)
+                               script@(MultiNodeScript l) =
+  let trace = runSimTrace sim
+
+      evsCMT :: Trace (SimResult ())
+                        (ConnectionManagerTrace
+                          SimAddr
+                          (ConnectionHandlerTrace
+                            UnversionedProtocol
+                            DataFlowProtocolData))
+      evsCMT = traceWithNameTraceEvents trace
+
+      upperBound = multiNodeScriptToCounters dataFlow l
+
+  in tabulate "ConnectionEvents" (map showCEvs l)
+    . counterexample (ppScript script)
+    . counterexample (Trace.ppTrace show show evsCMT)
+    . getAllProperty
+    . bifoldMap
+       ( \ case
+           MainReturn {} -> mempty
+           v             -> AllProperty
+                            $ counterexample (show v) (property False)
+       )
+       ( \ trs
+        -> case trs of
+          TrConnectionManagerCounters cmc ->
+            AllProperty
+              $ counterexample
+                  ("Upper bound is: " ++ show upperBound
+                  ++ "\n But got: " ++ show cmc)
+                  (property $ cmc <= upperBound)
+          _                               ->
+            mempty
+       )
+    $ evsCMT
+  where
+    -- We count all connections as prunable because we do not have a better way
+    -- to know what transitions will a given connection go through. We also
+    -- count every client connection as unidirectional because they do not
+    -- negotiate duplex data flow.  When closing outbound connections we need to
+    -- increase inboundConnections by one since it might trigger a
+    -- DemotedToColdDuplex transition.
+    multiNodeScriptToCounters :: DataFlow
+                              -> [ConnectionEvent Int TestAddr]
+                              -> ConnectionManagerCounters
+    multiNodeScriptToCounters df =
+      let ifDuplex = bool 0 1 (df == Duplex)
+          ifUni = bool 0 1 (df == Unidirectional)
+       in
+        foldMap' id
+        . foldl'
+          (\ st ce -> case ce of
+            StartClient _ ta ->
+              Map.insert ta (ConnectionManagerCounters 1 0 1 0 0) st
+            StartServer _ ta _ ->
+              Map.insert ta (ConnectionManagerCounters 1 ifDuplex ifUni 0 0) st
+            InboundConnection _ ta ->
+              Map.adjust (ConnectionManagerCounters 0 0 0 1 0 <>) ta st
+            OutboundConnection _ ta ->
+              Map.adjust (ConnectionManagerCounters 0 0 0 0 1 <>) ta st
+            InboundMiniprotocols _ ta _ ->
+              Map.adjust (ConnectionManagerCounters 0 0 0 0 0 <>) ta st
+            OutboundMiniprotocols _ ta _ ->
+              Map.adjust (ConnectionManagerCounters 0 0 0 0 0 <>) ta st
+            CloseOutboundConnection _ ta ->
+              Map.adjust (ConnectionManagerCounters 0 0 0 1 0 <>) ta st
+            _ -> st
+          )
+          Map.empty
```

Comment:
What's your thought about using information available at `Simulation.Network.Snocket.NetworkState`?  I think it should be posilbe to improve this test: rather than checking for an uperbound, we should be able to compute what's the current number of inbound / outbound connections.   Then we could use a special `Tracer` which would verify that the counters agree with the `NetworkState`.

Reply:
ok, ping me when you'll be ready with this.